openedx/core/djangoapps/config_model_utils/models.py

```diff
@@ -0,0 +1,125 @@
+
+from collections import namedtuple
+
+from django.conf import settings
+from django.db import models
+from django.contrib.sites.models import Site
+from django.utils.translation import ugettext_lazy as _
+
+from config_models.models import ConfigurationModel
+from openedx.core.djangoapps.site_configuration.models import SiteConfiguration
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+
+
+class StackedConfigurationModel(ConfigurationModel):
+    """
+    A ConfigurationModel that stacks Global, Site, Org, and Course level
+    configuration values.
+    """
+    class Meta(object):
+        abstract = True
+
+    KEY_FIELDS = ('site', 'org', 'course')
+
+    enabled = models.NullBooleanField(default=None, verbose_name=_("Enabled"))
+    site = models.ForeignKey(Site, on_delete=models.CASCADE, null=True)
+    org = models.CharField(max_length=255, db_index=True, null=True)
+    course = models.ForeignKey(
+        CourseOverview,
+        on_delete=models.DO_NOTHING,
+        null=True,
+    )
+
+    @classmethod
+    def attribute_tuple(cls):
+        "Returns a namedtuple with all attributes that can be overridden on this config model."
+        if hasattr(cls, '_attribute_tuple'):
+            return cls._attribute_tuple
+
+        cls._attribute_tuple = namedtuple(
+            '{}Values'.format(cls.__name__),
+            cls.STACKABLE_FIELDS
+        )
+        return cls._attribute_tuple
+
+    @classmethod
+    def current(cls, site=None, org=None, course=None):  # disable: pylint=arguments-differ
+        """
+        Return the current overridden configuration at the specified level.
+
+        Only one level may be specified at a time. Specifying multiple levels
+        will result in a ValueError.
+
+        Arguments:
+            site: The Site to check current values for
+            org: The org to check current values for
+            course: The course to check current values for
+
+        Returns:
+            An instance of :class:`cls.attribute_tuple()` with the overridden values
+            specified down to the level of the supplied argument (or global values if
+            no arguments are supplied).
+        """
+        if len([arg for arg in [site, org, course] if arg is not None]) > 1:
+            raise ValueError("Only one of site, org, and course can be specified")
+
+        if org is None and course is not None:
+            org = cls._org_from_course(course)
+
+        if site is None and org is not None:
+            site = cls._site_from_org(org)
+
+        stackable_fields = [cls._meta.get_field(field_name) for field_name in cls.STACKABLE_FIELDS]
+        field_defaults = {
+            field.name: field.get_default()
+            for field in stackable_fields
+        }
+
+        values = field_defaults.copy()
+
+        global_current = super(StackedConfigurationModel, cls).current(None, None, None)
+        for field in stackable_fields:
```

Comment:
do you think it would be clearer if you created values and field_defaults as part of one loop?  
then you won't need to copy or do two loops.

Reply:
The copy is cheap (it's a shallow copy). Not sure if the intent is clearer by assigning twice in a loop or by using copy.